Describe Git for Windows' architecture

dscho opened the pull request:
The Git for Windows project has grown quite complex over the years, certainly much more complex than during the first years where the msysgit.git repository was abusing Git for package management purposes and the git/git fork was called 4msysgit.git.

Let's describe the status quo in a thorough way.

rimrul commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,116 @@
+
+Git for Windows is a complex project.
+
+First and foremost, it is a friendly fork of [https://github.com/git/git](`git/git`), aiming to improve Git's Windows support. The [https://github.com/git-for-windows/git](`git-for-windows/git`) repository contains dozens of topics on top of `git/git`, some awaiting to be "upstreamed" (i.e. to be contributed to `git/git`), some still being stabilized, and a few topics are specific to the Git for  ...
```
Markdown link syntax is the other way around. 

```markdown
[Description](https://example.com)
```

dscho replied:
D'oh. Thank you!

rimrul commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,116 @@
+
+Another aspect that contributes to the complexity of Git for Windows is that it is not just building `git.exe` and distributes that. Due to its heritage within the Linux project, Git takes certain things for granted, such as the presence of a Unix shell, or for that matter, a package management system from which dependencies can be fetched and updated independently of Git itself. Things that are  ...
+
+Most of MSYS2's packages that are bundled in Git for Windows are consumed directly from MSYS2. Others need forks that are maintained by Git for Windows project, to support Git for Windows better. These forks live in the [`git-for-windows/MSYS2-packages`](https://github.com/git-for-windows/MSYS2-packages) and [`git-for-windows/MINGW-packages`](https://github.com/git-for-windows/MINGW-packages) rep ...
```
/but longer any/but no longer builds any/

rimrul commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,116 @@
+
+Most of MSYS2's packages that are bundled in Git for Windows are consumed directly from MSYS2. Others need forks that are maintained by Git for Windows project, to support Git for Windows better. These forks live in the [`git-for-windows/MSYS2-packages`](https://github.com/git-for-windows/MSYS2-packages) and [`git-for-windows/MINGW-packages`](https://github.com/git-for-windows/MINGW-packages) rep ...
+
+Since the infrastructure required by Git is non-trivial the installer (or for that matter, the Portable Git) is not exactly light-weight: As of January 2023, both artifacts are weigh over fifty megabytes. This is a problem for third-party applications wishing to bundle a version of Git for Windows, which is often advisable given that applications may depend on features that have been introduced o ...
```
/are weigh/are/

vdye commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,116 @@
+
+In contrast to the Git project, Git for Windows not only publishes tagged versions, but full builds of Git. In fact, Git for Windows' primary purpose, as far as most users are concerned, is to provide a convenient installer that end-users can run to have Git on their computer, without ever having to check out `git-for-windows/git` let alone build it. In essence, Git for Windows has to maintain a  ...
+
+Another aspect that contributes to the complexity of Git for Windows is that it is not just building `git.exe` and distributes that. Due to its heritage within the Linux project, Git takes certain things for granted, such as the presence of a Unix shell, or for that matter, a package management system from which dependencies can be fetched and updated independently of Git itself. Things that are  ...
```
s/`As Git for Windows 2.x`/`As of Git for Windows 2.x`?

vdye commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,116 @@
+
+Most of MSYS2's packages that are bundled in Git for Windows are consumed directly from MSYS2. Others need forks that are maintained by Git for Windows project, to support Git for Windows better. These forks live in the [`git-for-windows/MSYS2-packages`](https://github.com/git-for-windows/MSYS2-packages) and [`git-for-windows/MINGW-packages`](https://github.com/git-for-windows/MINGW-packages) rep ...
+
+Since the infrastructure required by Git is non-trivial the installer (or for that matter, the Portable Git) is not exactly light-weight: As of January 2023, both artifacts are weigh over fifty megabytes. This is a problem for third-party applications wishing to bundle a version of Git for Windows, which is often advisable given that applications may depend on features that have been introduced o ...
```
I'm not sure I understand what you mean by "non-interactive parts". Does that include `git.exe`?

dscho replied:
I changed it to "[...] contains only the parts of Git for Windows relevant for third-party applications. It lacks Git GUI, for example, as well as the terminal program MinTTY, or for that matter, the documentation."

vdye commented on ARCHITECTURE.md:
````diff
@@ -0,0 +1,116 @@
+
+Here is a diagram showing how the bits and pieces fit together.
+
+```mermaid
+graph LR
+  A[`monitor-components`] --> |opens| B
+  B{issues labeled<br />`component-update`} --> |/open pr| C
+  C((GitForWindowsHelper)) --> |triggers| D
+  D[`open-pr`] --> |opens| E
+  E{PR in</br>MINGW-packages<br />MSYS2-packages<br />build-extra} --> |closes| B
+  E --> |/deploy| F
+  F((GitForWindowsHelper)) --> |triggers| G
+  G[`build-and-deploy`] --> |deploys to| H
+  H{Pacman repository}
+  C --> |backed by| I
+  F --> |backed by| I
+  I[[Azure Function]]
+  D --> |running in| J
+  G --> | running in| J
+  J[[git-for-windows-automation]]
+  K[[git-sdk-32<br />git-sdk-64<br />git-sdk-arm64]] --> |syncing from| H
+  B --> |/add release note| L
+  L[`add-release-note`]
+```
````
This diagram is great! 🧜‍♀️ It makes this whole setup (and the remainder of your documentation here) much easier to follow.

vdye reviewed (approved):
This is beautifully written; it captures a lot of complex information while still feeling easy to learn/internalize. I just had a minor typo/nit & a question about MinGit. Otherwise, :shipit: !

dscho commented:
I believe that I addressed all the concerns that were raised in this ticket. In addition, I added some headers in the first part, to break up what was a big wall of text.

@rimrul @vdye could I ask you to have another quick look?

rimrul commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,130 @@
+
+### Providing more than just source code
+
+In contrast to the Git project, Git for Windows not only publishes tagged source code versions, but full builds of Git. In fact, Git for Windows' primary purpose, as far as most users are concerned, is to provide a convenient installer that end-users can run to have Git on their computer, without ever having to check out `git-for-windows/git` let alone build it. In essence, Git for Windows has to ...
```
The build-extra link is also flipped.

rimrul commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,130 @@
+
+### Supporting third-party applications that use Git's functionality
+
+Since the infrastructure required by Git is non-trivial the installer (or for that matter, the Portable Git) is not exactly light-weight: As of January 2023, both artifacts are over fifty megabytes. This is a problem for third-party applications wishing to bundle a version of Git for Windows, which is often advisable given that applications may depend on features that have been introduced only in ...
```
/provides MinGit as release artifact/provides MinGit as a release artifact

rimrul commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,130 @@
+
+### Deploying the Pacman packages
+
+This will open a Pull Request in one of Git for Windows' repository, and once the PR build passes, a Git for Windows maintainer issues the `/deploy` command ([example](https://github.com/git-for-windows/MINGW-packages/pull/69#issuecomment-1427591890)), which gets picked up by the GitForWindowsHelper GitHub App, which triggers [the `build-and-deploy` workflow](https://github.com/git-for-windows/gi ...
```
/one of Git for Windows' repository/one of Git for Windows' repositories/

rimrul reviewed (commented):
There's still some minor issues, but otherwise it looks good to me.
